README.md
[![Yii2](https://img.shields.io/badge/required-Yii2_v2.0.40-blue.svg)](https://packagist.org/packages/yiisoft/yii2)
[![Downloads](https://img.shields.io/packagist/dt/wdmg/yii2-sitemap.svg)](https://packagist.org/packages/wdmg/yii2-sitemap)
[![Packagist Version](https://img.shields.io/packagist/v/wdmg/yii2-sitemap.svg)](https://packagist.org/packages/wdmg/yii2-sitemap)
![Progress](https://img.shields.io/badge/progress-ready_to_use-green.svg)
[![GitHub license](https://img.shields.io/github/license/wdmg/yii2-sitemap.svg)](https://github.com/wdmg/yii2-sitemap/blob/master/LICENSE)

<img src="./docs/images/yii2-sitemap.png" width="100%" alt="Yii2 Sitemap" />

# Yii2 Sitemap
The module polls supported data models and displays links to their representations for front-ends and search engines.

You can expand or replace the list of supported models through the module configuration.
In your model class, which you include in the site map module, you need to provide a selection by publication status (for example), as well data models must have the required public attribute `in_sitemap` and public method` getPublished() `which should return models with such required keys as: `url`,` updated_at`.

Autogenerated sitemap resolving at http://example.com/sitemap.xml

Sitemap module monitors the saving, changing or deleting event for the supported models and clean the cache which the site map is formed, thus actualizing it.
In addition, you can add/edit and delete the URL of sitemap in manual mode.

> Be careful when adding URLs manually for pages of dynamic content, since by deleting such a resource the link to it will still be on the site map.

This module is an integral part of the [Butterfly.СMS](https://butterflycms.com/) content management system, but can also be used as an standalone extension.

Copyrights (c) 2019-2023 [W.D.M.Group, Ukraine](https://wdmg.com.ua/)

# Requirements 
* PHP 5.6 or higher
* Yii2 v.2.0.40 and newest
* [Yii2 Base](https://github.com/wdmg/yii2-base) module (required)
* [Yii2 SelectInput](https://github.com/wdmg/yii2-selectinput) widget
* [Yii2 Options](https://github.com/wdmg/yii2-options) module (optionality)
* [Yii2 Pages](https://github.com/wdmg/yii2-pages) module (support)
* [Yii2 News](https://github.com/wdmg/yii2-news) module (support)
* [Yii2 Blog](https://github.com/wdmg/yii2-blog) module (support)

# Installation
To install the module, run the following command in the console:

`$ composer require "wdmg/yii2-sitemap"`

After configure db connection, run the following command in the console:

`$ php yii sitemap/init`

And select the operation you want to perform:
  1) Apply all module migrations
  2) Revert all module migrations
  3) Flush sitemap cache

# Migrations
In any case, you can execute the migration and create the initial data, run the following command in the console:

`$ php yii migrate --migrationPath=@vendor/wdmg/yii2-sitemap/migrations`

# Configure
To add a module to the project, add the following data in your configuration file:

    'modules' => [
        ...
        'sitemap' => [
            'class' => 'wdmg\sitemap\Module',
            'routePrefix' => 'admin',
            'supportModels'  => [  // list of supported models for displaying a sitemap
                'pages' => 'wdmg\pages\models\Pages',
                'news' => 'wdmg\news\models\News',
            ],
            'cacheExpire' => 43200, // sitemap cache lifetime, `0` - for not use cache
            'defaultFrequency' => 'weekly', // default update frequency
            'defaultPriority' => 0.5, // default update priority
            'sitemapRoute' => '/' // default route to rendered sitemap.xml (use "/" - for root)
        ],
        ...
    ],


# Routing
Use the `Module::dashboardNavItems()` method of the module to generate a navigation items list for NavBar, like this:

    <?php
        echo Nav::widget([
        'options' => ['class' => 'navbar-nav navbar-right'],
            'label' => 'Modules',
            'items' => [
                Yii::$app->getModule('sitemap')->dashboardNavItems(),
                ...
            ]
        ]);
    ?>

# Status and version [ready to use]
* v.1.2.0 - Update copyrights, fix nav menu
* v.1.1.4 - Update dependencies, README.md
* v.1.1.3 - Update README.md and dependencies


CHANGELOG.md
Changelog
=========

## 1.2.0 (2023-06-29)
 * Update copyrights, fix nav menu

## 1.1.4 (2020-06-28)
 * Update dependencies, README.md

## 1.1.3 (2020-04-28)
 * Update README.md and dependencies
 
## 1.1.2 (2020-04-17)
 * Clear the runtime cache
 * Added support for Blog module, fixed models items retrieved

## 1.1.1 (2020-03-25)
 * Added pagination
 * Html encode for url`s in xml
 
## 1.1.0 (2020-01-09)
 * Added validation rules for URL`s
 * Added CRUD for edit sitemap URL`s
 
## 1.0.0 (2020-01-05)
 * Added support for Pages and News models
 * Added console, migrations and controller
 
## 0.0.1 (2019-12-20)
 * Initial commit

Module.php
<?php

namespace wdmg\sitemap;

/**
 * Yii2 Sitemap manager
 *
 * @category        Module
 * @version         1.2.0
 * @author          Alexsander Vyshnyvetskyy <alex.vyshnyvetskyy@gmail.com>
 * @link            https://github.com/wdmg/yii2-sitemap
 * @copyright       Copyright (c) 2019 - 2023 W.D.M.Group, Ukraine
 * @license         https://opensource.org/licenses/MIT Massachusetts Institute of Technology (MIT) License
 *
 */

use wdmg\helpers\ArrayHelper;
use Yii;
use wdmg\base\BaseModule;
use yii\base\InvalidConfigException;
use yii\helpers\Url;

/**
 * Sitemap module definition class
 */
class Module extends BaseModule
{
    /**
     * {@inheritdoc}
     */
    public $controllerNamespace = 'wdmg\sitemap\controllers';

    /**
     * {@inheritdoc}
     */
    public $defaultRoute = "list/index";

    /**
     * @var string, the name of module
     */
    public $name = "Sitemap";

    /**
     * @var string, the description of module
     */
    public $description = "Sitemap manager";

    /**
     * @var array list of supported models for displaying a sitemap
     */
    public $supportModels = [
        'pages' => 'wdmg\pages\models\Pages',
        'news' => 'wdmg\news\models\News',
        'blog' => 'wdmg\blog\models\Posts'
    ];

    /**
     * @var int sitemap cache lifetime, `0` - for not use cache
     */
    public $cacheExpire = 43200;

    /**
     * @var string default update frequency for sitemap.xml items
     * See https://www.sitemaps.org/protocol.html#xmlTagDefinitions
     */
    public $defaultFrequency = 'weekly';

    /**
     * @var float default update priority for sitemap.xml items
     * See https://www.sitemaps.org/protocol.html#xmlTagDefinitions
     */
    public $defaultPriority = 0.5;

    /**
     * @var string default route to rendered sitemap.xml (use "/" - for root)
     */
    public $sitemapRoute = "/";

    /**
     * @var string the module version
     */
    private $version = "1.2.0";

    /**
     * @var integer, priority of initialization
     */
    private $priority = 4;

    /**
     * {@inheritdoc}
     */
    public function init()
    {
        parent::init();

        // Set version of current module
        $this->setVersion($this->version);

        // Set priority of current module
        $this->setPriority($this->priority);

        // Process and normalize route for sitemap in frontend
        $this->sitemapRoute = self::normalizeRoute($this->sitemapRoute);

    }

    /**
     * {@inheritdoc}
     */
    public function dashboardNavItems($options = null)
    {
        $items = [
            'label' => $this->name,
            'icon' => 'fa fa-fw fa-sitemap',
            'url' => [$this->routePrefix . '/'. $this->id],
            'active' => (in_array(\Yii::$app->controller->module->id, [$this->id]) &&  Yii::$app->controller->id == 'list'),
        ];

	    if (!is_null($options)) {

		    if (isset($options['count'])) {
			    $items['label'] .= '<span class="badge badge-default float-right">' . $options['count'] . '</span>';
			    unset($options['count']);
		    }

		    if (is_array($options))
			    $items = ArrayHelper::merge($items, $options);

	    }

	    return $items;
    }

    /**
     * {@inheritdoc}
     */
    public function bootstrap($app)
    {
        parent::bootstrap($app);

        if (isset(Yii::$app->params["sitemap.supportModels"]))
            $this->supportModels = Yii::$app->params["sitemap.supportModels"];

        if (isset(Yii::$app->params["sitemap.cacheExpire"]))
            $this->cacheExpire = Yii::$app->params["sitemap.cacheExpire"];

        if (isset(Yii::$app->params["sitemap.defaultFrequency"]))
            $this->defaultFrequency = Yii::$app->params["sitemap.defaultFrequency"];

        if (isset(Yii::$app->params["sitemap.defaultPriority"]))
            $this->defaultPriority = Yii::$app->params["sitemap.defaultPriority"];

        if (isset(Yii::$app->params["sitemap.sitemapRoute"]))
            $this->sitemapRoute = Yii::$app->params["sitemap.sitemapRoute"];

        if (!isset($this->supportModels))
            throw new InvalidConfigException("Required module property `supportModels` isn't set.");

        if (!isset($this->cacheExpire))
            throw new InvalidConfigException("Required module property `cacheExpire` isn't set.");

        if (!isset($this->defaultFrequency))
            throw new InvalidConfigException("Required module property `defaultFrequency` isn't set.");

        if (!isset($this->defaultPriority))
            throw new InvalidConfigException("Required module property `defaultPriority` isn't set.");

        if (!isset($this->sitemapRoute))
            throw new InvalidConfigException("Required module property `sitemapRoute` isn't set.");

        if (!is_array($this->supportModels))
            throw new InvalidConfigException("Module property `supportModels` must be array.");

        if (!is_integer($this->cacheExpire))
            throw new InvalidConfigException("Module property `cacheExpire` must be integer.");

        if (!is_string($this->defaultFrequency))
            throw new InvalidConfigException("Module property `defaultFrequency` must be a string.");

        if (!in_array($this->defaultFrequency, ['always', 'hourly', 'daily', 'weekly', 'monthly', 'yearly', 'never']))
            throw new InvalidConfigException("Module property `defaultFrequency` must be one of the values: always, hourly, daily, weekly, monthly, yearly, never.");

        if (!is_float($this->defaultPriority))
            throw new InvalidConfigException("Module property `defaultPriority` must be float integer.");

        if (!is_string($this->sitemapRoute))
            throw new InvalidConfigException("Module property `sitemapRoute` must be a string.");

        // Add route to pass sitemap.xml in frontend
        $sitemapRoute = $this->sitemapRoute;
        if (empty($sitemapRoute) || $sitemapRoute == "/") {
            $app->getUrlManager()->addRules([
                [
                    'pattern' => '/sitemap',
                    'route' => 'admin/sitemap/default',
                    'suffix' => '.xml'
                ],
                '/sitemap.xml' => 'admin/sitemap/default'
            ], true);
        } else if (is_string($sitemapRoute)) {
            $app->getUrlManager()->addRules([
                [
                    'pattern' => $sitemapRoute . '/sitemap',
                    'route' => 'admin/sitemap/default',
                    'suffix' => '.xml'
                ],
                $sitemapRoute . '/sitemap.xml' => 'admin/sitemap/default'
            ], true);
        }

        // Attach to events of create/change/remove of models for the subsequent clearing cache of sitemap
        if (!($app instanceof \yii\console\Application)) {
            if (is_array($models = $this->supportModels) && $cache = $app->getCache()) {
                foreach ($models as $name => $class) {
                    if (class_exists($class)) {
                        $model = new $class();
                        \yii\base\Event::on($class, $model::EVENT_AFTER_INSERT, function ($event) use ($cache) {
                            $cache->delete(md5('sitemap'));
                        });
                        \yii\base\Event::on($class, $model::EVENT_AFTER_UPDATE, function ($event) use ($cache) {
                            $cache->delete(md5('sitemap'));
                        });
                        \yii\base\Event::on($class, $model::EVENT_AFTER_DELETE, function ($event) use ($cache) {
                            $cache->delete(md5('sitemap'));
                        });
                    }
                }
            }
        }
    }

    /**
     * Generate current sitemap.xml URL
     *
     * @return null|string
     */
    public function getSitemapURL() {
        $url = null;
        $sitemapRoute = $this->sitemapRoute;
        if (empty($sitemapRoute) || $sitemapRoute == "/") {
            $url = Url::to('/sitemap.xml', true);
        } else {
            $url = Url::to($sitemapRoute . '/sitemap.xml', true);
        }
        return $url;
    }
}